// appium-base-driver/build/lib/jsonwp-proxy/protocol-converter.js
'use strict';

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _defineProperty = require('babel-runtime/helpers/define-property')['default'];

var _slicedToArray = require('babel-runtime/helpers/sliced-to-array')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _basedriverDriver = require('../basedriver/driver');

var _basedriverDriver2 = _interopRequireDefault(_basedriverDriver);

var _appiumSupport = require('appium-support');

var log = _appiumSupport.logger.getLogger('Protocol Converter');

var COMMAND_URLS_CONFLICTS = [{
  commandNames: ['execute', 'executeAsync'],
  jsonwpConverter: function jsonwpConverter(url) {
    return url.replace(/\/execute.*/, url.includes('async') ? '/execute_async' : '/execute');
  },
  w3cConverter: function w3cConverter(url) {
    return url.replace(/\/execute.*/, url.includes('async') ? '/execute/async' : '/execute/sync');
  }
}, {
  commandNames: ['getElementScreenshot'],
  jsonwpConverter: function jsonwpConverter(url) {
    return url.replace(/\/element\/([^\/]+)\/screenshot$/, '/screenshot/$1');
  },
  w3cConverter: function w3cConverter(url) {
    return url.replace(/\/screenshot\/([^\/]+)/, '/element/$1/screenshot');
  }
}, {
  commandNames: ['getWindowHandles', 'getWindowHandle'],
  jsonwpConverter: function jsonwpConverter(url) {
    return url.replace(/\/window\/handle(s?)$/, '/window_handle$1');
  },
  w3cConverter: function w3cConverter(url) {
    return url.replace(/\/window_handle(s?)$/, '/window/handle$1');
  }
}];

var _BaseDriver$DRIVER_PROTOCOL = _basedriverDriver2['default'].DRIVER_PROTOCOL;
var MJSONWP = _BaseDriver$DRIVER_PROTOCOL.MJSONWP;
var W3C = _BaseDriver$DRIVER_PROTOCOL.W3C;

var ProtocolConverter = (function () {
  function ProtocolConverter(proxyFunc) {
    _classCallCheck(this, ProtocolConverter);

    this.proxyFunc = proxyFunc;
    this._downstreamProtocol = null;
  }

  _createClass(ProtocolConverter, [{
    key: 'getTimeoutRequestObjects',

    /**
     * W3C /timeouts can take as many as 3 timeout types at once, MJSONWP /timeouts only takes one
     * at a time. So if we're using W3C and proxying to MJSONWP and there's more than one timeout type
     * provided in the request, we need to do 3 proxies and combine the result
     *
     * @param {Object} body Request body
     * @return {Array} Array of W3C + MJSONWP compatible timeout objects
     */
    value: function getTimeoutRequestObjects(body) {
      var typeToW3C, _ret;

      return _regeneratorRuntime.async(function getTimeoutRequestObjects$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (!(this.downstreamProtocol === W3C && _lodash2['default'].has(body, 'ms') && _lodash2['default'].has(body, 'type'))) {
              context$2$0.next = 3;
              break;
            }

            typeToW3C = function typeToW3C(x) {
              return x === 'page load' ? 'pageLoad' : x;
            };

            return context$2$0.abrupt('return', [_defineProperty({}, typeToW3C(body.type), body.ms)]);

          case 3:
            if (!(this.downstreamProtocol === MJSONWP && (!_lodash2['default'].has(body, 'ms') || !_lodash2['default'].has(body, 'type')))) {
              context$2$0.next = 7;
              break;
            }

            _ret = (function () {
              var typeToJSONWP = function typeToJSONWP(x) {
                return x === 'pageLoad' ? 'page load' : x;
              };
              return {
                v: _lodash2['default'].toPairs(body).filter(function (pair) {
                  return !isNaN(parseFloat(pair[1]));
                }).map(function (pair) {
                  return {
                    type: typeToJSONWP(pair[0]),
                    ms: pair[1]
                  };
                })
              };
            })();

            if (!(typeof _ret === 'object')) {
              context$2$0.next = 7;
              break;
            }

            return context$2$0.abrupt('return', _ret.v);

          case 7:
            return context$2$0.abrupt('return', [body]);

          case 8:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }

    /**
     * Proxy an array of timeout objects and merge the result
     * @param {String} url Endpoint url
     * @param {String} method Endpoint method
     * @param {Object} body Request body
     */
  }, {
    key: 'proxySetTimeouts',
    value: function proxySetTimeouts(url, method, body) {
      var response, resBody, timeoutRequestObjects, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, timeoutObj, _ref2, _ref22;

      return _regeneratorRuntime.async(function proxySetTimeouts$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            response = undefined, resBody = undefined;
            context$2$0.next = 3;
            return _regeneratorRuntime.awrap(this.getTimeoutRequestObjects(body));

          case 3:
            timeoutRequestObjects = context$2$0.sent;

            log.debug('Will send the following request bodies to /timeouts: ' + JSON.stringify(timeoutRequestObjects));
            _iteratorNormalCompletion = true;
            _didIteratorError = false;
            _iteratorError = undefined;
            context$2$0.prev = 8;
            _iterator = _getIterator(timeoutRequestObjects);

          case 10:
            if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
              context$2$0.next = 25;
              break;
            }

            timeoutObj = _step.value;
            context$2$0.next = 14;
            return _regeneratorRuntime.awrap(this.proxyFunc(url, method, timeoutObj));

          case 14:
            _ref2 = context$2$0.sent;
            _ref22 = _slicedToArray(_ref2, 2);
            response = _ref22[0];
            resBody = _ref22[1];

            if (!(this.downstreamProtocol !== MJSONWP)) {
              context$2$0.next = 20;
              break;
            }

            return context$2$0.abrupt('return', [response, resBody]);

          case 20:
            if (!(response.statusCode >= 400)) {
              context$2$0.next = 22;
              break;
            }

            return context$2$0.abrupt('return', [response, resBody]);

          case 22:
            _iteratorNormalCompletion = true;
            context$2$0.next = 10;
            break;

          case 25:
            context$2$0.next = 31;
            break;

          case 27:
            context$2$0.prev = 27;
            context$2$0.t0 = context$2$0['catch'](8);
            _didIteratorError = true;
            _iteratorError = context$2$0.t0;

          case 31:
            context$2$0.prev = 31;
            context$2$0.prev = 32;

            if (!_iteratorNormalCompletion && _iterator['return']) {
              _iterator['return']();
            }

          case 34:
            context$2$0.prev = 34;

            if (!_didIteratorError) {
              context$2$0.next = 37;
              break;
            }

            throw _iteratorError;

          case 37:
            return context$2$0.finish(34);

          case 38:
            return context$2$0.finish(31);

          case 39:
            return context$2$0.abrupt('return', [response, resBody]);

          case 40:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[8, 27, 31, 39], [32,, 34, 38]]);
    }
  }, {
    key: 'proxySetWindow',
    value: function proxySetWindow(url, method, body) {
      var bodyObj;
      return _regeneratorRuntime.async(function proxySetWindow$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            bodyObj = _appiumSupport.util.safeJsonParse(body);

            if (!_lodash2['default'].isPlainObject(bodyObj)) {
              context$2$0.next = 12;
              break;
            }

            if (!(this.downstreamProtocol === W3C && _lodash2['default'].has(bodyObj, 'name') && !_lodash2['default'].has(bodyObj, 'handle'))) {
              context$2$0.next = 7;
              break;
            }

            log.debug('Reassigned \'name\' value \'' + bodyObj.name + '\' to \'handle\' as per W3C spec');
            context$2$0.next = 6;
            return _regeneratorRuntime.awrap(this.proxyFunc(url, method, { handle: bodyObj.name }));

          case 6:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 7:
            if (!(this.downstreamProtocol === MJSONWP && _lodash2['default'].has(bodyObj, 'handle') && !_lodash2['default'].has(bodyObj, 'name'))) {
              context$2$0.next = 12;
              break;
            }

            log.debug('Reassigned \'handle\' value \'' + bodyObj.handle + '\' to \'name\' as per JSONWP spec');
            context$2$0.next = 11;
            return _regeneratorRuntime.awrap(this.proxyFunc(url, method, { name: bodyObj.handle }));

          case 11:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 12:
            context$2$0.next = 14;
            return _regeneratorRuntime.awrap(this.proxyFunc(url, method, body));

          case 14:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 15:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }

    /**
     * Handle "crossing" endpoints for the case
     * when upstream and downstream drivers operate different protocols
     *
     * @param {string} commandName
     * @param {string} url
     * @param {string} method
     * @param {?string|object} body
     * @returns The proxyfying result as [response, responseBody] tuple
     */
  }, {
    key: 'convertAndProxy',
    value: function convertAndProxy(commandName, url, method, body) {
      var _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, _step2$value,

      // Same arguments, but different URLs
      commandNames, jsonwpConverter, w3cConverter, rewrittenUrl;

      return _regeneratorRuntime.async(function convertAndProxy$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (this.downstreamProtocol) {
              context$2$0.next = 4;
              break;
            }

            context$2$0.next = 3;
            return _regeneratorRuntime.awrap(this.proxyFunc(url, method, body));

          case 3:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 4:
            context$2$0.t0 = commandName;
            context$2$0.next = context$2$0.t0 === 'timeouts' ? 7 : context$2$0.t0 === 'setWindow' ? 10 : 13;
            break;

          case 7:
            context$2$0.next = 9;
            return _regeneratorRuntime.awrap(this.proxySetTimeouts(url, method, body));

          case 9:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 10:
            context$2$0.next = 12;
            return _regeneratorRuntime.awrap(this.proxySetWindow(url, method, body));

          case 12:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 13:
            return context$2$0.abrupt('break', 14);

          case 14:
            _iteratorNormalCompletion2 = true;
            _didIteratorError2 = false;
            _iteratorError2 = undefined;
            context$2$0.prev = 17;
            _iterator2 = _getIterator(COMMAND_URLS_CONFLICTS);

          case 19:
            if (_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done) {
              context$2$0.next = 37;
              break;
            }

            _step2$value = _step2.value;
            commandNames = _step2$value.commandNames;
            jsonwpConverter = _step2$value.jsonwpConverter;
            w3cConverter = _step2$value.w3cConverter;

            if (commandNames.includes(commandName)) {
              context$2$0.next = 26;
              break;
            }

            return context$2$0.abrupt('continue', 34);

          case 26:
            rewrittenUrl = this.downstreamProtocol === MJSONWP ? jsonwpConverter(url) : w3cConverter(url);

            if (!(rewrittenUrl === url)) {
              context$2$0.next = 30;
              break;
            }

            log.debug('Did not know how to rewrite the original URL \'' + url + '\' ' + ('for ' + this.downstreamProtocol + ' protocol'));
            return context$2$0.abrupt('break', 37);

          case 30:
            log.info('Rewrote the original URL \'' + url + '\' to \'' + rewrittenUrl + '\' ' + ('for ' + this.downstreamProtocol + ' protocol'));
            context$2$0.next = 33;
            return _regeneratorRuntime.awrap(this.proxyFunc(rewrittenUrl, method, body));

          case 33:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 34:
            _iteratorNormalCompletion2 = true;
            context$2$0.next = 19;
            break;

          case 37:
            context$2$0.next = 43;
            break;

          case 39:
            context$2$0.prev = 39;
            context$2$0.t1 = context$2$0['catch'](17);
            _didIteratorError2 = true;
            _iteratorError2 = context$2$0.t1;

          case 43:
            context$2$0.prev = 43;
            context$2$0.prev = 44;

            if (!_iteratorNormalCompletion2 && _iterator2['return']) {
              _iterator2['return']();
            }

          case 46:
            context$2$0.prev = 46;

            if (!_didIteratorError2) {
              context$2$0.next = 49;
              break;
            }

            throw _iteratorError2;

          case 49:
            return context$2$0.finish(46);

          case 50:
            return context$2$0.finish(43);

          case 51:
            context$2$0.next = 53;
            return _regeneratorRuntime.awrap(this.proxyFunc(url, method, body));

          case 53:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 54:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[17, 39, 43, 51], [44,, 46, 50]]);
    }
  }, {
    key: 'downstreamProtocol',
    set: function set(value) {
      this._downstreamProtocol = value;
    },
    get: function get() {
      return this._downstreamProtocol;
    }
  }]);

  return ProtocolConverter;
})();

exports['default'] = ProtocolConverter;
module.exports = exports['default'];

// If we got a non-MJSONWP response, return the result, nothing left to do

// If we got an error, return the error right away

// ...Otherwise, continue to the next timeouts call

// There is no point to convert anything if we do not know
// for which protocol the conversion should be done

// Same url, but different arguments

// No matches found. Proceed normally
