// appium-base-driver/build/test/protocol/fake-driver.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _2 = require('../..');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _uuidJs = require('uuid-js');

var _uuidJs2 = _interopRequireDefault(_uuidJs);

var FakeDriver = (function (_BaseDriver) {
  _inherits(FakeDriver, _BaseDriver);

  function FakeDriver() {
    _classCallCheck(this, FakeDriver);

    _get(Object.getPrototypeOf(FakeDriver.prototype), 'constructor', this).call(this);
    this.protocol = _2.BaseDriver.DRIVER_PROTOCOL.MJSONWP;
    this.sessionId = null;
    this.jwpProxyActive = false;
  }

  _createClass(FakeDriver, [{
    key: 'sessionExists',
    value: function sessionExists(sessionId) {
      if (!sessionId) {
        return false;
      }
      return sessionId === this.sessionId;
    }
  }, {
    key: 'driverForSession',
    value: function driverForSession() /*sessionId*/{
      return this;
    }
  }, {
    key: 'createSession',
    value: function createSession(desiredCapabilities, requiredCapabilities, capabilities) {
      return _regeneratorRuntime.async(function createSession$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            // Use a counter to make sure each session has a unique id
            this.sessionId = 'fakeSession_' + _uuidJs2['default'].create().hex;

            if (!capabilities) {
              context$2$0.next = 5;
              break;
            }

            return context$2$0.abrupt('return', [this.sessionId, capabilities]);

          case 5:
            this.desiredCapabilities = desiredCapabilities;
            this.requiredCapabilities = requiredCapabilities || {};
            return context$2$0.abrupt('return', [this.sessionId, _lodash2['default'].extend({}, desiredCapabilities, requiredCapabilities)]);

          case 8:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'executeCommand',
    value: function executeCommand(cmd) {
      for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        args[_key - 1] = arguments[_key];
      }

      return _regeneratorRuntime.async(function executeCommand$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (this[cmd]) {
              context$2$0.next = 2;
              break;
            }

            throw new _2.errors.NotYetImplementedError();

          case 2:
            if (cmd === 'createSession') {
              this.protocol = _2.BaseDriver.determineProtocol.apply(null, args);
            }
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap(this[cmd].apply(this, args));

          case 5:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 6:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'deleteSession',
    value: function deleteSession() {
      return _regeneratorRuntime.async(function deleteSession$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            this.jwpProxyActive = false;
            this.sessionId = null;

          case 2:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'getStatus',
    value: function getStatus() {
      return _regeneratorRuntime.async(function getStatus$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', "I'm fine");

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'setUrl',
    value: function setUrl(url) {
      return _regeneratorRuntime.async(function setUrl$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', 'Navigated to: ' + url);

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'getUrl',
    value: function getUrl() {
      return _regeneratorRuntime.async(function getUrl$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', "http://foobar.com");

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'back',
    value: function back(sessionId) {
      return _regeneratorRuntime.async(function back$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', sessionId);

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'forward',
    value: function forward() {
      return _regeneratorRuntime.async(function forward$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'refresh',
    value: function refresh() {
      return _regeneratorRuntime.async(function refresh$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            throw new Error('Too Fresh!');

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'getSession',
    value: function getSession() {
      return _regeneratorRuntime.async(function getSession$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            throw new _2.errors.NoSuchDriverError();

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'click',
    value: function click(elementId, sessionId) {
      return _regeneratorRuntime.async(function click$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', [elementId, sessionId]);

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'implicitWait',
    value: function implicitWait(ms) {
      return _regeneratorRuntime.async(function implicitWait$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', ms);

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'clickCurrent',
    value: function clickCurrent(button) {
      return _regeneratorRuntime.async(function clickCurrent$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', button);

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'setNetworkConnection',
    value: function setNetworkConnection(type) {
      return _regeneratorRuntime.async(function setNetworkConnection$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', type);

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'moveTo',
    value: function moveTo(element, xOffset, yOffset) {
      return _regeneratorRuntime.async(function moveTo$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', [element, xOffset, yOffset]);

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'getText',
    value: function getText() {
      return _regeneratorRuntime.async(function getText$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', "");

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'getAttribute',
    value: function getAttribute(attr, elementId, sessionId) {
      return _regeneratorRuntime.async(function getAttribute$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', [attr, elementId, sessionId]);

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'setValue',
    value: function setValue(value, elementId) {
      return _regeneratorRuntime.async(function setValue$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', [value, elementId]);

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'performTouch',
    value: function performTouch() {
      for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        args[_key2] = arguments[_key2];
      }

      return _regeneratorRuntime.async(function performTouch$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', args);

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'setFrame',
    value: function setFrame(frameId) {
      return _regeneratorRuntime.async(function setFrame$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', frameId);

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'removeApp',
    value: function removeApp(app) {
      return _regeneratorRuntime.async(function removeApp$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', app);

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'receiveAsyncResponse',
    value: function receiveAsyncResponse() {
      return _regeneratorRuntime.async(function receiveAsyncResponse$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', { status: 13, value: 'Mishandled Driver Error' });

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'proxyActive',
    value: function proxyActive() /*sessionId*/{
      return false;
    }
  }, {
    key: 'getProxyAvoidList',
    value: function getProxyAvoidList() /*sessionId*/{
      return [];
    }
  }, {
    key: 'canProxy',
    value: function canProxy() /*sessionId*/{
      return false;
    }
  }]);

  return FakeDriver;
})(_2.BaseDriver);

exports.FakeDriver = FakeDriver;

// this is here to test a failing command that does not throw an error
